Fixes #36099 - Fix search options method for lifecycle environments

ofedoren opened the pull request.

wbclark reviewed (commented):
Hey @ofedoren , this came up in Katello triage which brought it to my attention.

I am wondering if you can provide some additional context which I couldn't find in the commit message, redmine issue (maybe it comes from some RH bugzilla but I don't see any attached)

1. what was the error that occurred when trying `hammer hostgroup create/update --name <name> --puppet-environment <name> --lifecycle-environment <name> --query-organization-id 1` with the puppet plugin installed?

2. what was the issue with the current option order / why does reversing the order fix it / is there any potential for introducing a similar type of issue with the new order?

Sorry if these are rather basic questions, as I am not a hammer expert

ofedoren commented:
> 1. what was the error that occurred when trying `hammer hostgroup create/update --name <name> --puppet-environment <name> --lifecycle-environment <name> --query-organization-id 1` with the puppet plugin installed?

My bad, I'll update the issue with debug output and insert here as well. TL;DR: The problem is that name->id resolution for  `--puppet-enviornment` option goes through lifecycle environment resource endpoint instead of Puppet plugin's one.

Error:
```bash
hammer -d hostgroup update --name test --puppet-environment dev --lifecycle-environment Library --query-organization-id 1
```
```ruby
[ INFO 2023-02-16T13:56:09 API] GET /katello/api/organizations/1/environments
[DEBUG 2023-02-16T13:56:09 API] Params: {
       "name" => "dev",
    :per_page => 1000,
        :page => 1
}
[DEBUG 2023-02-16T13:56:09 API] Headers: {
    :params => {
           "name" => "dev",
        :per_page => 1000,
            :page => 1
    }
}
[DEBUG 2023-02-16T13:56:09 API] Using authenticator: HammerCLIForeman::Api::InteractiveBasicAuth
[DEBUG 2023-02-16T13:56:10 API] Response: {
         "total" => 0,
      "subtotal" => 0,
    "selectable" => 0,
          "page" => "1",
      "per_page" => "1000",
         "error" => nil,
        "search" => nil,
          "sort" => {
           "by" => "name",
        "order" => "asc"
    },
       "results" => []
}
[DEBUG 2023-02-16T13:56:10 API] Response headers: {
                                 :date => "Thu, 16 Feb 2023 13:56:09 GMT",
                               :server => "Apache",
                      :foreman_version => "3.6.0-develop",
                  :foreman_api_version => "2",
         :foreman_current_organization => "1; Default Organization",
             :foreman_current_location => "; ANY",
                         :content_type => "application/json; charset=utf-8",
                      :apipie_checksum => "6fb70fc1dd918536eac2f0f265c3abddc6136aa2",
                                 :etag => "W/\"2a2540031476365e734ade9c75785f31\"",
                        :cache_control => "max-age=0, private, must-revalidate",
                         :x_request_id => "33a7b6a3-469a-4be7-9d3d-eb3b6601142c",
                            :x_runtime => "0.304501",
            :strict_transport_security => "max-age=631139040; includeSubdomains",
                      :x_frame_options => "sameorigin",
               :x_content_type_options => "nosniff",
                     :x_xss_protection => "1; mode=block",
                   :x_download_options => "noopen",
    :x_permitted_cross_domain_policies => "none",
              :content_security_policy => "default-src 'self'; child-src 'self'; connect-src 'self' ws: wss:; img-src 'self' data:; script-src 'unsafe-eval' 'unsafe-inline' 'self'; style-src 'unsafe-inline' 'self'",
                           :set_cookie => [
        [0] "_session_id=1234bfb6bfc54ddbe411d9d7fb1ab9cb; path=/; HttpOnly; secure; SameSite=Lax"
    ],
                                  :via => "1.1 foreman.helo.com",
                    :transfer_encoding => "chunked"
}
[DEBUG 2023-02-16T13:56:10 Exception] Using exception handler HammerCLIForeman::ExceptionHandler#handle_general_exception
[ERROR 2023-02-16T13:56:10 Exception] Error: lifecycle_environment not found.
Could not update the hostgroup:
  Error: lifecycle_environment not found.
[ERROR 2023-02-16T13:56:10 Exception] 

HammerCLIForeman::ResolverError (lifecycle_environment not found.):
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/id_resolver.rb:250:in `pick_result'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/id_resolver.rb:191:in `find_resource'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/id_resolver.rb:154:in `get_id'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/id_resolver.rb:142:in `block (2 levels) in define_id_finders'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/commands.rb:142:in `get_resource_id'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/option_sources/id_params.rb:48:in `block in get_options'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/option_sources/id_params.rb:43:in `each'
    /home/vagrant/hammer-cli-foreman/lib/hammer_cli_foreman/option_sources/id_params.rb:43:in `get_options'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/sources/base.rb:8:in `process'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:22:in `block in process'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:21:in `each'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:21:in `inject'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:21:in `process'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:22:in `block in process'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:21:in `each'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:21:in `inject'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/processor_list.rb:21:in `process'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/option_collector.rb:17:in `all_options_raw'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/option_collector.rb:21:in `all_options'
    /home/vagrant/hammer-cli/lib/hammer_cli/options/option_collector.rb:25:in `options'
    /home/vagrant/hammer-cli/lib/hammer_cli/abstract.rb:389:in `options'
    /home/vagrant/hammer-cli/lib/hammer_cli/abstract.rb:117:in `parse'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:62:in `run'
    /home/vagrant/hammer-cli/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /home/vagrant/hammer-cli/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /home/vagrant/hammer-cli/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
    /home/vagrant/hammer-cli/bin/hammer:147:in `<top (required)>'
    /usr/share/gems/bin/hammer:23:in `load'
    /usr/share/gems/bin/hammer:23:in `<top (required)>'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/cli/exec.rb:63:in `load'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/cli/exec.rb:63:in `kernel_load'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/cli/exec.rb:28:in `run'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/cli.rb:475:in `exec'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/cli.rb:31:in `dispatch'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/cli.rb:25:in `start'
    /usr/share/gems/gems/bundler-2.2.24/libexec/bundle:46:in `block in <top (required)>'
    /usr/share/gems/gems/bundler-2.2.24/lib/bundler/friendly_errors.rb:128:in `with_friendly_errors'
    /usr/share/gems/gems/bundler-2.2.24/libexec/bundle:34:in `<top (required)>'
    /usr/bin/bundle:23:in `load'
    /usr/bin/bundle:23:in `<main>'
```

Note that `dev` is not a lifecycle environment and thus should not resolve via Katello API, but Puppet.

> 2. what was the issue with the current option order / why does reversing the order fix it / is there any potential for introducing a similar type of issue with the new order?

I've added similar change a long time ago, but probably I didn't test all combinations or I simply overlooked the bug. The real issue is that previously we had `--environment` options that could imply either `lifecycle` or `puppet`, hammer couldn't figure the difference while building options from API docs, thus a lot of workarounds were added until all the such options will be deprecated and removed later. Some options were removed already, but some commands still have deprecated ones. The custom `create_lifecycle_environments_search_options` search option creator looks at provided option values and if I'm not mistaken due to compatibility the value can be stored as either `option_name` or `option_environment_name`. In case Puppet is installed, `--puppet-environment` get stored into `option_environment_name` which can be mistreated by `create_lifecycle_environments_search_options`, thus the method should look for `option_name` at first place.

It would be quite awesome if someone will test other commands that have such options, not necessary a command that contains both envs, but rather one (lifecycle for instance). Activation keys command maybe?.. Not sure where comething could fail.

> Sorry if these are rather basic questions, as I am not a hammer expert

That's totally OK, I guess I'm the only one left who deals with hammer projects :)

chris1984 reviewed (approved):
ACK
